<a href="https://colab.research.google.com/github/plajuan/ML_Study/blob/main/Data_Preparation_Cookbook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cookbook

## Data preparation

Data preparation is used in machine learning in order to make the data ready for the application of algorithms. As such, data preparation is a processing step in which the original data (*raw data*) are manipulated in order to clean, organize and transform them, so that analysis and machine learning algorithm processes can be executed.

In this cookbook, we'll look at a set of common data preparation techniques. The illustrated techniques use the [pandas](https://pandas.pydata.org/) library.

## Problem 1

Create a new *DataFrame*.

In [4]:
import pandas as pd
df_e = pd.DataFrame(columns=["col1", 'col2'], data=[['Areia', 'Mineral'], ['Madeira', 'Vegetal']])
df_e.loc[len(df_e)] = ['Pedra', 'Mineral']
df_e

col1     col2
0    Areia  Mineral
1  Madeira  Vegetal
2    Pedra  Mineral

### Solution

In [ ]:
# Import library
import pandas as pd

# Create dataframe
df = pd.DataFrame()

# Add columns to the dataframe
df['Name'] = ['Joana', 'Miguel', 'Carla']
df['Age'] = [19, 21, 22]

# Show dataframe
df

Name  Age
0   Joana   19
1  Miguel   21
2   Carla   22

### Comments

There are several ways to create *DataFrames* using pandas. The example illustrated above is just one way. In practice, the procedure used above is not the most common. The most common is to import data from any data source (e.g., database or CSV file).

## Problem 2

View some features of the *DataFrame*.



In [5]:
df_e.head()

col1     col2
0    Areia  Mineral
1  Madeira  Vegetal
2    Pedra  Mineral

### Solução

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Show the first 5 rows
df.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
# Show the dimensions of the dataframe
df.shape

(891, 12)

In [ ]:
# Create a statistical summary
df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

### Comments

After importing the data, it's a good idea to understand how it is structured and what information it contains. Ideally we would see all the data, however, as we often have thousands of observations, it is necessary to use the above techniques to see only parts of the data.

In the example above, we use the Titanic database as a reference. We displayed the first lines and learned that it has 891 observations (people) and 12 variables for each person (characteristics). Note that, in the end, the summary statistics do not include some variables (e.g., 'Name') because these are numerical.

## Problem 3

Access only parts of the *DataFrame*.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Select the first row
df.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                                 22
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

In [ ]:
# Select information from the second, third and fourth row
df.iloc[1:4]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S

[3 rows x 12 columns]

In [ ]:
# Select information from all rows up to the fourth
df.iloc[:4]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S

[4 rows x 12 columns]

In [ ]:
df = df.set_index('Name')
df

PassengerId  ...  Embarked
Name                                                             ...          
Braund, Mr. Owen Harris                                       1  ...         S
Cumings, Mrs. John Bradley (Florence Briggs Tha...            2  ...         C
Heikkinen, Miss. Laina                                        3  ...         S
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  4  ...         S
Allen, Mr. William Henry                                      5  ...         S
...                                                         ...  ...       ...
Montvila, Rev. Juozas                                       887  ...         S
Graham, Miss. Margaret Edith                                888  ...         S
Johnston, Miss. Catherine Helen "Carrie"                    889  ...         S
Behr, Mr. Karl Howell                                       890  ...         C
Dooley, Mr. Patrick                                         891  ...         Q

[891 rows x 11 columns]

In [ ]:
df.loc['Braund, Mr. Owen Harris']

PassengerId            1
Survived               0
Pclass                 3
Sex                 male
Age                   22
SibSp                  1
Parch                  0
Ticket         A/5 21171
Fare                7.25
Cabin                NaN
Embarked               S
Name: Braund, Mr. Owen Harris, dtype: object

### Comments
All lines in the *DataFrame* have an ID. `iloc` allows us to filter the information based on the position of the row in the *DataFrame*. In turn, `loc` filters information based on labels (*labels*) that are associated with the row (e.g., name).

## Problem 4

Select rows based on conditions.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Select the first two lines where 'Sex' == 'female'
df[df['Sex']=='female'].head(2)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S

[2 rows x 12 columns]

In [ ]:
# Select the observations referring to females over the age of 50
df[(df['Sex']=='female') & (df['Age']>50)]

PassengerId  Survived  Pclass  ...      Fare Cabin  Embarked
11            12         1       1  ...   26.5500  C103         S
15            16         1       2  ...   16.0000   NaN         S
195          196         1       1  ...  146.5208   B80         C
268          269         1       1  ...  153.4625  C125         S
275          276         1       1  ...   77.9583    D7         S
366          367         1       1  ...   75.2500   D37         C
483          484         1       3  ...    9.5875   NaN         S
496          497         1       1  ...   78.2667   D20         C
513          514         1       1  ...   59.4000   NaN         C
571          572         1       1  ...   51.4792  C101         S
591          592         1       1  ...   78.2667   D20         C
765          766         1       1  ...   77.9583   D11         S
772          773         0       2  ...   10.5000   E77         S
774          775         1       2  ...   23.0000   NaN         S
820          821         1       1  ...   93.5000   B69         S
829          830         1       1  ...   80.0000   B28       NaN
879          880         1       1  ...   83.1583   C50         C

[17 rows x 12 columns]

### Comments

Filtering information based on conditions is one of the most common tasks in data preparation. Typically, we are interested in parts of the information rather than the whole.

## Problem 5

Change *DataFrame* values.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Change values from 'female' to 'Woman'
df['Sex'].replace('female', 'Woman').head()

0     male
1    Woman
2    Woman
3    Woman
4     male
Name: Sex, dtype: object

In [ ]:
# Change values from 'female' and 'male' to 'Woman' and 'Man'
df['Sex'].replace(['female', 'male'], ['Woman', 'Man']).head()

0      Man
1    Woman
2    Woman
3    Woman
4      Man
Name: Sex, dtype: object

### Comments

The `replace` function is used when data transformations are needed. Sometimes, in order to make our data compatible, these transformations are necessary.

## Problem 6

Change the column names of a *DataFrame*.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Change the name of the columns
df.rename(columns={'Name':'NewName','Pclass':'Class','Age':'NewAge'})

PassengerId  Survived  Class  ...     Fare Cabin  Embarked
0              1         0      3  ...   7.2500   NaN         S
1              2         1      1  ...  71.2833   C85         C
2              3         1      3  ...   7.9250   NaN         S
3              4         1      1  ...  53.1000  C123         S
4              5         0      3  ...   8.0500   NaN         S
..           ...       ...    ...  ...      ...   ...       ...
886          887         0      2  ...  13.0000   NaN         S
887          888         1      1  ...  30.0000   B42         S
888          889         0      3  ...  23.4500   NaN         S
889          890         1      1  ...  30.0000  C148         C
890          891         0      3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

### Comment

Using `rename` in conjunction with a dictionary is one of the simplest ways to change column names. In certain cases, changing the column names helps us understand the data.

## Problem 7

Calculate the maximum, minimum, mean, and sum of a numeric column.


### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Calculate Basic Statistics for the 'Age' Column
print('Maximum: {}'.format(df['Age'].max()))
print('Minimum: {}'.format(df['Age'].min()))
print('Mean: {}'.format(df['Age'].mean()))
print('Sum: {}'.format(df['Age'].sum()))

Maximum: 80.0
Minimum: 0.42
Mean: 29.69911764705882
Sum: 21205.17


### Comments

Basic statistical methods are already programmed in the *pandas* library. So to use them, just call the respective methods. In addition to these methods, there are others that are also useful, such as `var`, `std` and `count`.

## Problem 8

Find all unique values of a date column.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Identify the unique values in a column
df['Pclass'].unique()

array([3, 1, 2])

In [ ]:
# Identify unique values and their frequency
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

## Problem 9

Identify the missing values in a given column.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Identify observations without 'Age' information
df[df['Age'].isnull()]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
5              6         0       3  ...   8.4583   NaN         Q
17            18         1       2  ...  13.0000   NaN         S
19            20         1       3  ...   7.2250   NaN         C
26            27         0       3  ...   7.2250   NaN         C
28            29         1       3  ...   7.8792   NaN         Q
..           ...       ...     ...  ...      ...   ...       ...
859          860         0       3  ...   7.2292   NaN         C
863          864         0       3  ...  69.5500   NaN         S
868          869         0       3  ...   9.5000   NaN         S
878          879         0       3  ...   7.8958   NaN         S
888          889         0       3  ...  23.4500   NaN         S

[177 rows x 12 columns]

In [ ]:
# Identify observations with 'Age' information
df[df['Age'].notnull()]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
885          886         0       3  ...  29.1250   NaN         Q
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[714 rows x 12 columns]

## Problem 10

Remove columns from the *DataFrame*.


### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Remove the columns 'Age' and 'PClass'
df.drop(['Age','Pclass'], axis=1)

PassengerId  Survived  ... Cabin Embarked
0              1         0  ...   NaN        S
1              2         1  ...   C85        C
2              3         1  ...   NaN        S
3              4         1  ...  C123        S
4              5         0  ...   NaN        S
..           ...       ...  ...   ...      ...
886          887         0  ...   NaN        S
887          888         1  ...   B42        S
888          889         0  ...   NaN        S
889          890         1  ...  C148        C
890          891         0  ...   NaN        Q

[891 rows x 10 columns]

### Comments

There are other ways to eliminate columns, but this one is the most common. When columns are dropped and we want to make this action permanent, we either activate the `inplace` parameter by setting `inplace=True`, or we store the *DataFrame* with the deleted column in a variable.

## Problem 11

Check the dataset for missing data.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Check if there is missing data
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Problem 12

Remove observations with missing data.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Eliminate observations with missing data
df = df.dropna()
df

PassengerId  Survived  Pclass  ...     Fare        Cabin  Embarked
1              2         1       1  ...  71.2833          C85         C
3              4         1       1  ...  53.1000         C123         S
6              7         0       1  ...  51.8625          E46         S
10            11         1       3  ...  16.7000           G6         S
11            12         1       1  ...  26.5500         C103         S
..           ...       ...     ...  ...      ...          ...       ...
871          872         1       1  ...  52.5542          D35         S
872          873         0       1  ...   5.0000  B51 B53 B55         S
879          880         1       1  ...  83.1583          C50         C
887          888         1       1  ...  30.0000          B42         S
889          890         1       1  ...  30.0000         C148         C

[183 rows x 12 columns]

## Problem 13

Assign all missing values using average values.

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Assign missing values using average values
df = df.fillna(df.mean())
df

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

## Problem 14

Assign missing data to a specific column (using the mean).

### Solution

In [ ]:
# Import library
import pandas as pd

# Create URL
url = 'https://raw.githubusercontent.com/pmarcelino/datasets/master/titanic.csv'

# Import data
df = pd.read_csv(url)

# Assign missing data to the column 'Age' using average values
df['Age'] = df['Age'].fillna(df['Age'].mean())
df

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]